### PR TITLE
Correctly handle literal \ufffd chars when encoding JSON

### LINKED ISSUE #313: 1.1.7 encode crash
{"code":0, "data":{"text":"�ɪV�+|�!��^r�Ǘ�\u0026\b\u0005˵U۳.\f���","texts":[{"source":"v1Vg24nSA19fK0b67rX1YGToxXbv8mz52VNfFus3TDU7A=","target":"�ɪV�+|�!��^r�Ǘ�\u0026\b\u0005˵U۳.\f���"}]}}

when i decode this string to map[string]interface{}, it work fine. but encode to json string, the encoder use all of my memory, the app crashed

### LINKED ISSUE #317: Freeze when encoding unicode replacement character
Hello. 

Probably found json encoding bug for strings containing unicode replacement character.

```
package main

import (
	"fmt"
	"unicode/utf8"

	"github.com/ugorji/go/codec"
)

func main() {
	var err error

	// Unicode REPLACEMENT CHARACTER U+FFFD
	a := []byte{0xef, 0xbf, 0xbd}
	as := string(a)

	m := map[string]interface{}{
		"somekey1": as,
	}

	// is valid rune
	fmt.Println(utf8.Valid(a))
	// decodes OK
	fmt.Println(utf8.DecodeRune(a))

	jhandle := &codec.JsonHandle{}

	buf := make([]byte, 0, 1024)
	jenc := codec.NewEncoderBytes(&buf, jhandle)
	// 100% CPU forever
	err = jenc.Encode(&m)
	if err != nil {
		panic(err)
	}

	// never go here
}
```

### PR DESCRIPTION
**Pull request by cvermilion**
Fixes #313.
Fixes #317.

utf8.DecodeRuneInString uses this rune to signal errors, but it can also be present as a literal rune in the string.

Without this fix, encoding a string that contains `�` results in an infinite loop. I wasn't sure where the best place to extend the tests would be; adding a string with this rune in it to the `primitives` list at the top causes tests involving other encoders to fail.

### CODE REVIEW

File: codec/json.go [hunk 1]
```diff
@@ -412,15 +412,18 @@ func (e *jsonEncDriver) quoteStr(s string) {
...
+			// c=utf8.RuneError, size=0: empty input (shouldn't happen here since i < len(s))
+			// c=utf8.RuneError, size=1: invalid encoding (handle below by skipping a byte and writing the replacement character \ufffd
+			// c=utf8.RuneError, size=2,4: shouldn't ever be returned
+			// c=utf8.RuneError, size=3: literal '\ufffd' in input, this is ok
+			if start < i {
+				w.writestr(s[start:i])
 			}
+			w.writestr(`\ufffd`)
+			i++
+			start = i
```
Comment (cvermilion) [inline codec/json.go]: The key fix is that before, the case of `c, size == utf8.RuneError, 3` wasn't handled and instead resulted in an infinite loop.

Comment (disconnect3d) [reply codec/json.go]: The fix and test looks okay for the replacement character case and invalid unicode code points.

But what about other unicode characters? How should it work? It seems that only line separator and paragraph separator cases are handled later on and if this does not match, the character is skipped? (`i += uint(size)`).

There is also a `TestJsonInvalidUnicode` but I don't see a test that would check for valid unicode characters marshalled into JSON.

Am I missing something?

Comment (disconnect3d) [reply codec/json.go]: Also https://github.com/ugorji/go/blob/a4365be44eb789f34688916df4ce68edc2541ae2/codec/codec_test.go#L3335 comment should probably be removed...

Comment (cvermilion) [reply codec/json.go]: Invalid unicode was already handled; that's the `c, size == utf8.RuneError, 1` case. What's added here is an actual `\ufffd` replacement character in the input, which is valid Unicode. And note that other unicode characters are handled similar to ASCII -- incrementing `i` without changing `start`. At the end of the loop, or before writing any output, this encoder flushes all the bytes from `start` to `i`.

I suspect you are right about that comment, I think it can go.

**Comment by 70l571y** [comment]
We suggest this solution:
``` GO
c, _ := utf8.DecodeRuneInString(s[i:]) //because len(utf8.RuneError) is always 3
if c == utf8.RuneError {
	if start < i {
		w.writestr(s[start:i])
	}
	w.writestr(`\ufffd`)
	i += uint(size)
	start = i
	continue
}
```
Its worked for us well
